@vadmium thanks for the review! Yes, the test will fail without the fix:

```
======================================================================
ERROR: testDontResetInternalStateOnException (test.test_wsgiref.HandlerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 138, in run
    self.finish_response()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 184, in finish_response
    self.write(data)
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 289, in write
    self.send_headers()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 347, in send_headers
    self.send_preamble()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 267, in send_preamble
    self._write(('HTTP/%s %s\r\n' % (self.http_version,self.status)).encode('iso-8859-1'))
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 468, in _write
    result = self.stdout.write(data)
  File "/Users/berkerpeksag/projects/cpython/Lib/test/test_wsgiref.py", line 819, in write
    raise CustomException
test.test_wsgiref.HandlerTests.testDontResetInternalStateOnException.<locals>.CustomException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/berkerpeksag/projects/cpython/Lib/test/test_wsgiref.py", line 824, in testDontResetInternalStateOnException
    h.run(hello_app)
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 145, in run
    self.handle_error()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 383, in handle_error
    self.finish_response()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 184, in finish_response
    self.write(data)
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 289, in write
    self.send_headers()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 346, in send_headers
    if not self.origin_server or self.client_is_modern():
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 359, in client_is_modern
    return self.environ['SERVER_PROTOCOL'].upper() != 'HTTP/0.9'
TypeError: 'NoneType' object is not subscriptable

----------------------------------------------------------------------
```

--- a/Lib/test/test_wsgiref.py
+++ b/Lib/test/test_wsgiref.py
@@ -806,6 +806,31 @@ class HandlerTests(TestCase):
 
                 self.assertFalse(stderr.getvalue())
 
+    def testDontResetInternalStateOnException(self):
+        class CustomException(ValueError):
+            pass
+
+        # We are raising CustomException here to trigger an exception
+        # during the execution of SimpleHandler.finish_response(), so
+        # we can easily test that the internal state of the handler is
+        # preserved in case of an exception.
+        class AbortingWriter:
+            def write(self, b):
+                raise CustomException
+
+        stderr = StringIO()
+        environ = {"SERVER_PROTOCOL": "HTTP/1.0"}
+        h = SimpleHandler(BytesIO(), AbortingWriter(), stderr, environ)
+        h.run(hello_app)
+
+        self.assertIn("CustomException", stderr.getvalue())
+
+        # Test that the internal state of the handler is preserved.
+        self.assertIsNotNone(h.result)
+        self.assertIsNotNone(h.headers)
+        self.assertIsNotNone(h.status)
+        self.assertIsNotNone(h.environ)
+
 
 if __name__ == "__main__":
     unittest.main()

--- a/Lib/wsgiref/handlers.py
+++ b/Lib/wsgiref/handlers.py
@@ -183,7 +183,16 @@ class BaseHandler:
                 for data in self.result:
                     self.write(data)
                 self.finish_content()
-        finally:
+        except:
+            # Call close() on the iterable returned by the WSGI application
+            # in case of an exception.
+            if hasattr(self.result, 'close'):
+                self.result.close()
+            raise
+        else:
+            # We only call close() when no exception is raised, because it
+            # will set status, result, headers, and environ fields to None.
+            # See bpo-29183 for more details.
             self.close()
 
 

--- a/Misc/NEWS.d/next/Library/2019-04-22-22-55-29.bpo-29183.MILvsk.rst
+++ b/Misc/NEWS.d/next/Library/2019-04-22-22-55-29.bpo-29183.MILvsk.rst
@@ -1,0 +1,3 @@
+Fix double exceptions in :class:`wsgiref.handlers.BaseHandler` by calling
+its :meth:`~wsgiref.handlers.BaseHandler.close` method only when no
+exception is raised.

